src/coreclr/jit/lower.cpp

```diff
@@ -7848,39 +7848,45 @@ void Lowering::ContainCheckBitCast(GenTree* node)
     }
 }
 
-struct StoreCoalescingData
+struct LoadStoreCoalescingData
 {
     var_types targetType;
     GenTree*  baseAddr;
     GenTree*  index;
     GenTree*  value;
     uint32_t  scale;
     int       offset;
+    bool      allowsNonAtomic;
 };
 
 //------------------------------------------------------------------------
-// GetStoreCoalescingData: given a STOREIND node, get the data needed to perform
+// GetLoadStoreCoalescingData: given a IND/STOREIND node, get the data needed to perform
 //    store coalescing including pointer to the previous node.
 //
 // Arguments:
 //    comp     - the compiler instance
-//    ind      - the STOREIND node
+//    ind      - the IND/STOREIND node
 //    data     - [OUT] the data needed for store coalescing
 //
 // Return Value:
 //    true if the data was successfully retrieved, false otherwise.
 //    Basically, false means that we definitely can't do store coalescing.
 //
-static bool GetStoreCoalescingData(Compiler* comp, GenTreeStoreInd* ind, StoreCoalescingData* data)
+static bool GetLoadStoreCoalescingData(Compiler* comp, GenTreeIndir* ind, LoadStoreCoalescingData* data)
 {
-    // Don't merge volatile stores.
+    assert(ind->OperIs(GT_IND, GT_STOREIND));
+
+    // Don't merge volatile stores/loads.
     if (ind->IsVolatile())
     {
         return false;
     }
 
-    // Data has to be INT_CNS, can be also VEC_CNS in future.
-    if (!ind->Data()->IsCnsIntOrI() && !ind->Data()->IsVectorConst())
+    const bool isStore = ind->OperIs(GT_STOREIND);
+
+    // Data is either a constant or GT_IND
+    // TODO-CoalescingStores: allow locals (to then broadcast them)
+    if (isStore && !ind->Data()->IsCnsIntOrI() && !ind->Data()->IsVectorConst() && !ind->Data()->OperIs(GT_IND))
     {
         return false;
     }
```

Comment:
This logic seems like it needs some form of interference checking somewhere when the data is `GT_IND`. What if the definition occurs 200 nodes before, with interfering stores?